### PR TITLE
FACT-407 Changes to opening time tab to make it a separate form that gets and posts its data via ajax. 

### PR DESCRIPTION
**Pull request by agraham01**
https://tools.hmcts.net/jira/browse/FACT-407

Changes to opening time tab to make it a separate form that gets and posts its data via ajax. The save functionality for the general tab is removed until the same changes are made to that tab.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/app/controller/courts/EditCourtGeneralController.ts
Comment (amstevenson) [inline src/main/app/controller/courts/EditCourtGeneralController.ts:12]: Is defaulting to null going to cause any issue? It looks like it gets set on line 16, so it isn't a huge problem if not. Is there a way to set a default empty value instead? 

Worried about null pointer exceptions, but this may not be a concern on the frontend, your call!

File: src/main/app/controller/courts/OpeningTimesController.ts
Comment (amstevenson) [inline src/main/app/controller/courts/OpeningTimesController.ts:54]: Can this be shortened to: `req.scope.cradle.api.updateOpeningTimes(req.params.slug, openingTimes)` ?

File: src/main/app/fact/FactApi.ts
Comment (amstevenson) [inline src/main/app/fact/FactApi.ts:8]: Seems that this file caters for a lot of different types of calls; admin opening, general, courts...types 

Is there a way to split this out into different files? I only mention this as it will grow and grow with time. If its not possible, then thats fair :) Just thought I'd mention

File: src/main/app/fact/FactApi.ts
Comment (amstevenson) [inline src/main/app/fact/FactApi.ts:56]: Just to double check on this one, the error will be raised back to the calling function on the controller, and from there the error message will be displayed. 

Will error be displayed to the user? And if not, will it be a general error that we have defined? If so, then great! If not, then we may want to log it in a debug sense to help us in the future

File: src/main/assets/js/opening-hours.ts
Comment (amstevenson) [inline src/main/assets/js/opening-hours.ts:9]: Are these methods private? 

Also, is there a naming convention for variables that are local to the scope of the file in javascript? 

For example (Global/Class Constants): 
java: TYPE_SELECT_NAME = 'type_id' < if at the top 
python: typically the same as java, but with _ at the start if private 

happy to take your steer on this one.

File: src/main/assets/js/opening-hours.ts
Comment (amstevenson) [inline src/main/assets/js/opening-hours.ts:47]: A small one, so ignore if its silly

1) The endpoint on the API is a put, and we make a call to it as a put in the src/main/assets/js/opening-hours.ts file. Is it worth declaring the other stuff as puts as well to make it more consistant?

File: src/main/assets/js/opening-hours.ts
Comment (amstevenson) [inline src/main/assets/js/opening-hours.ts:53]: Just to double check on this one if you wouldn't mind. 

If it fails to connect because the VPN isn't on, or because we are an invalid user, the console.log won't display the name of the IDAM location right? 

I sometimes see logs in the backend on a failure (of the main application) with stuff like "cant connect to blabla.idam.domain.thingy". Which might be bad if it also gets displayed on the frontend too as a log

File: src/main/modules/nunjucks/index.ts
Comment (amstevenson) [inline src/main/modules/nunjucks/index.ts:53]: Might be worth commenting why we are doing this here; It isn't too clear to me. 

It looks like if we have any element that is not empty, splice the data, but I don't understand why without looking into it further! Might help other developer in the future to add a quick note in

File: src/main/types/CourtPageData.ts
Comment (amstevenson) [inline src/main/types/CourtPageData.ts:10]: Is it possible to set this as a type rather than an `any` ? To make it a bit more strongly typed

File: src/main/types/OpeningTime.ts
Comment (amstevenson) [inline src/main/types/OpeningTime.ts:8]: Is it worth putting this into its own file? I.e OpeningType.ts ? Only a small one, not to worry if its a pain

File: src/test/functional/features/court-additional-information.feature
Comment (amstevenson) [inline src/test/functional/features/court-additional-information.feature:12]: Think these may have been commented out by accident?

File: src/test/functional/features/court-open-and-access-scheme.feature
Comment (amstevenson) [inline src/test/functional/features/court-open-and-access-scheme.feature:12]: same

File: src/test/functional/features/court-urgent.feature
Comment (amstevenson) [inline src/test/functional/features/court-urgent.feature:12]: same

File: src/test/functional/features/opening-hours.feature
Comment (amstevenson) [inline src/test/functional/features/opening-hours.feature:7]: I think Kian had a security vulnerability around exposing sensitive data. This would kind of fall under that, although it may not necessarily be picked up by the security team. 

Is it possible to use environment variables here to pick up these values, or something similar? Might be one we need to take offline and fix another time, since it's a known wider issue among the team that we all know about

File: src/test/functional/features/opening-hours.feature
Comment (amstevenson) [inline src/test/functional/features/opening-hours.feature:20]: Is it possible to check the new entry which has been added? And to check  that the text has been displayed as expected? I would imagine the step would need to be able to see the new element...so it would have to reload the page or something...

We can chat about this one if you want

File: src/test/functional/features/opening-hours.feature
Comment (amstevenson) [inline src/test/functional/features/opening-hours.feature:31]: Might be worth splitting the following into two for improved clarity: 

When('I click the remove button under an opening hours entry', async () => {
  const numOpeningTimes = await I.countElement('#openingTimesTab fieldset');

  const selector = 'button[name="deleteOpeningHours"]';
  const elementExist = await I.checkElement(selector);
  expect(elementExist).equal(true);
  await I.click(selector);

  const updatedNumOpeningTimes = await I.countElement('#openingTimesTab fieldset');
  expect(numOpeningTimes - updatedNumOpeningTimes).equal(1);
});

File: src/test/functional/steps/opening-hours.ts
Comment (amstevenson) [inline src/test/functional/steps/opening-hours.ts:9]: Can shorten to: 

`expect(await I.checkElement(selector)).equal(true);` if you want

File: src/test/functional/steps/opening-hours.ts
Comment (amstevenson) [inline src/test/functional/steps/opening-hours.ts:34]: Can this be reused for the other tabs? I.e set up with an element and a count parameter? Dw if its a pain

File: src/test/functional/utlis/puppeteer.util.ts
Comment (amstevenson) [inline src/test/functional/utlis/puppeteer.util.ts:130]: Optional one again, you may want to add a distinct message to help debug, like "console.log("The element didn't appear for the selectItem function with selector {} and value {}.");

File: src/main/app/controller/courts/OpeningTimesController.ts
Comment (agraham01) [reply src/main/app/controller/courts/OpeningTimesController.ts:12]: These are used in the unit tests for the controller. The tests use those messages as part of the expectation that a suitable error message will be displayed when the gets fail.

File: src/main/app/fact/FactApi.ts
Comment (agraham01) [reply src/main/app/fact/FactApi.ts:8]: This is a fair point and something I thought too. I wonder if we should refactor this class in a separate ticket. The existing methods all do pretty much the same thing (only the url changes) so could be improved. Error handling could be improved too. We'd need to test all endpoints if we refactor, hence I'm thinking we deal with this separately.

File: src/main/app/fact/FactApi.ts
Comment (agraham01) [reply src/main/app/fact/FactApi.ts:56]: The call to the existing error handler logs the error. Returning the rejected promise allows us to catch this error when calling the API method. In the case of opening times, we don't display the specific error returned in the UI but we do catch the error and display a meaningful message, such as: 'A problem occurred when retrieving the opening times.'

File: src/test/functional/features/court-open-and-access-scheme.feature
Comment (agraham01) [reply src/test/functional/features/court-open-and-access-scheme.feature:12]: See previous comment on other commented out functional tests.

File: src/test/functional/features/court-urgent.feature
Comment (agraham01) [reply src/test/functional/features/court-urgent.feature:12]: See previous comment on commented out functional tests.

File: src/test/functional/features/opening-hours.feature
Comment (agraham01) [reply src/test/functional/features/opening-hours.feature:7]: This exists in almost all of the other functional test files. It's also in the readme file and displays in github. I agree we should fix it but I think a separate ticket would be needed.

File: src/test/functional/steps/opening-hours.ts
Comment (agraham01) [reply src/test/functional/steps/opening-hours.ts:9]: I generally think async calls look a bit better on their own line.

File: src/main/app/controller/courts/EditCourtGeneralController.ts
Comment (agraham01) [reply src/main/app/controller/courts/EditCourtGeneralController.ts]: I've updated the code to remove the requirement for null.

File: src/main/app/controller/courts/OpeningTimesController.ts
Comment (agraham01) [reply src/main/app/controller/courts/OpeningTimesController.ts]: Code updated.

File: src/main/assets/js/opening-hours.ts
Comment (agraham01) [reply src/main/assets/js/opening-hours.ts]: I've refactored this into a proper typescript class and made these variables private.

File: src/main/assets/js/opening-hours.ts
Comment (agraham01) [reply src/main/assets/js/opening-hours.ts]: I think this is a good idea. The code has been updated to use PUT.

File: src/main/assets/js/opening-hours.ts
Comment (agraham01) [reply src/main/assets/js/opening-hours.ts]: I've updated the code and just removed the 'response' properties to avoid any such issues.

File: src/main/modules/nunjucks/index.ts
Comment (agraham01) [reply src/main/modules/nunjucks/index.ts]: Comment now added.

File: src/main/types/OpeningTime.ts
Comment (agraham01) [reply src/main/types/OpeningTime.ts]: Agreed - code updated with separate file for this type.

File: src/test/functional/steps/opening-hours.ts
Comment (agraham01) [reply src/test/functional/steps/opening-hours.ts:46]: I've refactored this into a re-usable method in the opening-hours.ts file. If there is a need for re-use, it can now be moved to a separate, common class easily.

File: src/test/functional/utlis/puppeteer.util.ts
Comment (agraham01) [reply src/test/functional/utlis/puppeteer.util.ts]: Agreed. Code updated.

File: src/test/functional/features/opening-hours.feature
Comment (Thomas-Geraghty) [reply src/test/functional/features/opening-hours.feature:7]: Hi gents, just on this security issue - for the API we’ve deployed the fixes utilising the AZ vault with terraform. We can deploy a similar fix here, as @agraham01 says it is definitely for another ticket though. We will need to get this addressed before go live.

**Review by Thomas-Geraghty — APPROVED** [review]
Looks great mate, agree with most of the comments already provided by @amstevenson.

File: src/test/functional/features/opening-hours.feature
Comment (agraham01) [reply src/test/functional/features/opening-hours.feature:31]: How do we get the number of expected items for use in the new 'Then' clause?

**Review by amstevenson — APPROVED** [review]

File: src/test/functional/features/opening-hours.feature
Comment (agraham01) [reply src/test/functional/features/opening-hours.feature:20]: Added a further step in updated code.

**Pull request merged**
